### PR TITLE
[DEV-639] - Add 'testUpgrade' hardhat task

### PR DESCRIPTION
**Pull request by grp06**
## Describe your changes
In the PR I added a `testUpgrade` task.

The benefit of this task is that as you develop your smart contracts, at any time you can run `npx hardhat testUpgrade` and quickly find out if your current changes will break upgradeability. In the current codebase there's no easy way to do this, since running upgrades requires an old ABI, which we're currently not keeping in source control.

When you run `npx hardhat testUpgrade` the following occurs
- we compile your contracts
- we deploy a proxy for `Raft`, `SpecDataHolder` and `Badges` contracts on your local hardhat network
- we take your most recent changes and attempt to run and upgrade against **the last release** from contracts (which is an ABI saved in `src/test/abis/latest`

In order to run the upgrade against **the last release** we must access the `abi` and `bytecode` of those contracts. 

In the future, as a separate ticket, we should create a Github action that automates this process. But for the time being, it'll be a manual job (that only takes a minute to copy/paste 3 files)

**When it's time for a release**, we can either 

- rename the`src/test/abis/latest` => `src/test/abis/${versionNumber}` (if we decide we want to preserve the ABIs for each release)
- or simply remove the contents of `src/test/abis/latest`

**Then** copy:
- `artifacts/src/Badges.sol/Badges.json` into `src/test/abis/latest`
- `artifacts/src/Raft.sol/Raft.json` into `src/test/abis/latest`
- `artifacts/src/SpecDataHolder.sol/SpecDataHolder.json` into `src/test/abis/latest`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tasks/testUpgrade.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+const specDataHolderAbi = specDataHolder.abi
+const specDataHolderBytecode = specDataHolder.bytecode
+
+import badges from '../test/abis/latest/Badges.json'
+const badgesAbi = badges.abi
+const badgesBytecode = badges.bytecode
+
+export default async function testUpgrade(params: any, hre: HardhatRuntimeEnvironment): Promise<void> {
+  // runs npx hardhat compile
+  // without this you'll the code below will be run with stale code
```
Comment (rahulrumalla) [inline tasks/testUpgrade.ts:20]: > // without this you'll the code below will be run with stale code

Can you update the comment to be more legible, please? I think it was a typo :)

File: tasks/testUpgrade.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  })
+  await raftProxy.deployed()
+
+  const specDataHolder = await ethers.getContractFactory(specDataHolderAbi, specDataHolderBytecode, owner)
+  const specDataHolderProxy = await upgrades.deployProxy(specDataHolder, [raftProxy.address, owner.address], {
+    kind: 'uups',
+  })
+  await specDataHolderProxy.deployed()
+
+  const name = 'Badges'
```
Comment (rahulrumalla) [inline tasks/testUpgrade.ts:38]: Can we read these from existing environment vars?

File: tasks/testUpgrade.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  await upgradedRaftContract.deployed()
+  console.log('RAFT contract upgraded successfully')
+
+  const specDataHolderImplementation = await ethers.getContractFactory('SpecDataHolder')
+  const upgradedspecDataHolderContract = await upgrades.upgradeProxy(
+    specDataHolderProxy.address,
+    specDataHolderImplementation
+  )
+  await upgradedspecDataHolderContract.deployed()
+  console.log('SPECDATAHOLDER contract upgraded successfully')
```
Comment (rahulrumalla) [inline tasks/testUpgrade.ts:68]: Should we write a few `expect` statements? or is the idea here to just ensure that there is a no-collision upgrade?

Comment (rahulrumalla) [reply tasks/testUpgrade.ts:68]: Looks like the tests that were deleted had some `expect` statements. Do share your thoughts behind not having these, please.

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
A few minor things. 

Also, do we need to update any readme instructions?

File: tasks/testUpgrade.ts (hunk 2 above)
Comment (grp06) [reply tasks/testUpgrade.ts]: done

File: tasks/testUpgrade.ts (hunk 1 above)
Comment (grp06) [reply tasks/testUpgrade.ts]: fixed ... comment not necessary anyway

**Review by rahulrumalla — APPROVED** [review]

**Pull request merged**